keps/sig-node/3953-node-resource-hot-plug/README.md

```diff
@@ -0,0 +1,704 @@
+
+With increase in cluster resources the following components will updated
+
+1. Scheduler
+   * Scheduler will automatically schedule any pending pods.
+
+
+2. Change in Swap Memory limit
+   * Currently, the swap memory limit is calculated as 
+ `(<containerMemoryRequest>/<nodeTotalMemory>)*<totalPodsSwapAvailable>`
+   * So increase in nodeTotalMemory will result in updated swap memory limit.
+
+
+3. Change in OOM score
+   * OOM score calculation depends on machine's memory, so the new OOM score will be updated accordingly.
+
+**Proposed Code changes**
+
+**Dynamic Node Scale Up logic**
+
+```go
+	if utilfeature.DefaultFeatureGate.Enabled(features.NodeResourceHotPlug) {
+		// Handle the node dynamic scale up
+		machineInfo, err := kl.cadvisor.MachineInfo()
+		if err != nil {
+			klog.ErrorS(err, "Error fetching machine info")
+		} else {
+			cachedMachineInfo, _ := kl.GetCachedMachineInfo()
+            // Avoid collector collects it as a timestamped metric
+            // See PR #95210 and #97006 for more details.
+                machineInfo.Timestamp = time.Time{}
+			if !reflect.DeepEqual(cachedMachineInfo, machineInfo) {
+				kl.setCachedMachineInfo(machineInfo)
+
+				// Resync the resource managers
+				if err := kl.ResyncComponents(machineInfo); err != nil {
+					klog.ErrorS(err, "Error resyncing the kubelet components with machine info")
+				}
+			}
+		}
+	}
+```
+
+**Changes to resource managers to adapt to dynamic scale up of resources**
+
+1. Adding ResyncComponents() method to ContainerManager interface
+```go
+    // Manages the containers running on a machine.
+    type ContainerManager interface {
+        .
+        .
+        // ResyncComponents will resyc the resource managers like cpu, memory and topology managers
+	// with updated machineInfo
+	ResyncComponents(machineInfo *cadvisorapi.MachineInfo) error
+	.
+	.
+    )
+```
+
+2. Adding a method Sync to all the resource managers and will be invoked once there is dynamic resource change.
+
+```go
+        // SyncMachineInfo will sync the Manager with the latest machine info
+        SyncMachineInfo(machineInfo *cadvisorapi.MachineInfo) error
+```
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Unit tests
+
+1. Add necessary tests in kubelet_node_status_test.go to check for the node status behaviour with dynamic node scale up.
+2. Add necessary tests in kubelet_pods_test.go to check for the pod cleanup and pod addition workflow.
+3. Add necessary tests in eventhandlers_test.go to check for scheduler behaviour with dynamic node capacity change.
+4. Add necessary tests in resource managers to check for managers behaviour to adopt dynamic node capacity change.
+
+
+##### e2e tests
+
+Following scenarios need to be covered:
+
+* Node resource information before and after resource hot plug.
```

Comment:
treating of `upsize -> downsize`  and `upsize -> downsize -> upsize` and `downsize` and `downsize->upsize` event sequence